zproject/email_backends.py

```diff
@@ -108,3 +110,46 @@ def send_messages(self, email_messages: Sequence[EmailMessage]) -> int:
                 email_log_url = settings.ROOT_DOMAIN_URI + "/emails"
                 logging.info("Emails sent in development are available at %s", email_log_url)
         return num_sent
+
+
+class PersistentSMTPEmailBackend(EmailBackend):
+    def _open(self, **kwargs: Any) -> bool | None:
+        is_opened = super().open()
+        if is_opened:
+            self.opened_at = timezone_now()
+            return True
+
+        return is_opened
+
+    @override
+    def open(self, **kwargs: Any) -> bool | None:
+        is_opened = self._open()
+        if is_opened:
+            return True
+
+        status = None
+        time_elapsed = (timezone_now() - self.opened_at).seconds / 60
+        if settings.EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES != 0 and (
+            settings.EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES is None
+            or time_elapsed <= settings.EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES
+        ):
+            # No-op to ensure that we don't return a connection that has been
+            # closed by the mail server.
+            try:
+                assert self.connection is not None
+                status = self.connection.noop()[0]
+            except Exception:
+                pass
+
+        # In case of EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES = 0, we
+        # will always return a newly opened connection. The default
+        # smtp.py implementation of send_messages closes the connection
+        # after sending the emails if the connection was a new
+        # connection i.e. is_opened is True.
```

Comment:
Doesn't this imply that we can never have a persistent connection?

That is, even if `EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES is None`, the _first_ time we run this, we're opening a new connection and thus returning `True`.  Which by this comment would imply that it will also get closed immediately after sending one message?

This comment, or a clarified version of it if accurate, is probably better at the top, inside the first `if is_opened`.

Reply:
OK, the solution to my mystery of two comments above is that:
1. If we're in the `email_senders` worker, we call `initialize_connection` which calls `open()`.
2. The default `send_messages` implementation calls `open()` again -- hence necessitating the check for `EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES != 0` above, since the `initialize_connection` means that even non-persistent connections will get `open()` called on them multiple times, which is quite non-intuitive.
3. So the above noop block is skipped for non-persistent connections, leaving `status = None`.
4. ...sooooo we hit the block below this and close the connection and re-open it, uselessly.
5. ...but that bug _also_ means that we return `True` from the second `open()` call, which means that `send_messages()` will then call `close()` for us.  Which is correct behaviour (albeit with the extra open/close)
6. ...but that _also also_ means that if we have a persistent connection, when we time out and hit the block below this, we'll return `True`, causing `send_messages()` to think it needs to clean up.  Which means it'll send one message over the newly-opened connection, and then close it.

So this logic is unfortunately still not correct.  I am beginning to suspect that we may want to _only use this subclass_ if persistent connections are used (`EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES != 0`) because it's not necessary in the case where it's `== 0`, and it complicates both the logic, and how easy the code is to reason about.